### DISCUSSION TITLE
Help for ODF edition

### DISCUSSION THREAD

**Discussion by eturpault** [GrandOrgue/grandorgue#917] [General]
Hello,
For whom it may interest, I share with you xml files that I made for Notepad++ editor to facilitate the edition of ODF (.organ) files :
- one which permits to highlight key words to make it more readable (file Notepad++ Language ODF.xml)
- one to have in the functions list of Notepad++ the list of the elements of the ODF to navigate more easily inside the file (file Notepad++ Functions ODF.xml)

ZIP file containing those two xml files : [Notepad++ xml for ODF.zip](https://github.com/GrandOrgue/grandorgue/files/7779894/Notepad%2B%2B.xml.for.ODF.zip)

Here is an example of what it looks like ;
![ODF Notepad++ screenshot](https://user-images.githubusercontent.com/51839171/147461256-17769b9a-715e-4b33-a9d7-955c641e6cb8.png)

To add the **language** file to Notepad++, go in the menu "Language", then select "Define your language..." at the bottom of the menu list. Inside the opened window, click on the button "Import..." and select the attached file "Notepad++ Language ODF.xml". An entry "ODF" has normally appeared in the top left dropdown list "User language". By selecting it you can see the properties of the xml file.
You can close the window a open an .orgal file to see the effect of the coloration.

To add the **functions list** file to Notepad++, copy the attached file "Notepad++ Functions ODF.xml" in your folder %APPDATA%\Notepad++\functionList, and rename it in "odf.xml". Restart Notepadd++ if it is already opened.
Enjoy it !

**Comment by jason-barba** [comment]
> To add the **functions list** file to Notepad++, copy the attached file "Notepad++ Functions ODF.xml" in your folder 
> %APPDATA%\Roaming\Notepad++\functionList, and rename it in "odf.xml". Restart Notepadd++ if it is already opened. Enjoy it !

This description is somewhat unclear to me. There is no %APPDATA%\Roaming\Notepad++\functionList directory on my disk, but there is a file functionList.xml. Should I include the content of Notepad++ Functions ODF.xml into functionList.xml? But if so, why rename the file?

**Reply by eturpault** [reply]
%APPDATA% is an environment variable of Windows which the value may depend on the configuration of your computer. To know which path it is you can open a DOS window (type and validate CMD in the search bar of Windows), then in the DOS window console type "%APPDATA%", it will show you the value of this variable. In my computer it is : "C:\Users\MyName\AppData\Roaming" (where MyName is my Windows user name). 
So the xml file has to be put in the following folder in my computer : C:\Users\MyName\AppData\Roaming\Notepad++\functionList
There is no need to include odf.xml inside the functionList.xml.
I rename the original file in odf.xml to make it more coherent with the other xml files of the folder.

**Reply by jason-barba** [reply]
> So the xml file has to be put in the following folder in my computer : C:\Users\MyName\AppData\Roaming\Notepad++\functionList There is no need to include odf.xml inside the functionList.xml. I rename the original file in odf.xml to make it more coherent with the other xml files of the folder.

So if the subfolder functionList doesn't exist, as in my case, it should just be created?

**Reply by jason-barba** [reply]
I created a folder **functionList** in %APPDATA%\Roaming\Notepad++, and moved the to **ODF.xml** renamed file **Notepad++ Functions ODF.xml** to it. Then i started Notepad++ and opened the Function List window, but it is empty.

What am I doing wrong?

**Reply by eturpault** [reply]
Information about the functions list feature of Notepad++ can be found here in the user manual : https://npp-user-manual.org/docs/function-list/
If you installed Notepad++ as a portable application, then the functionList folder is located in Notepad++ installation directory.

I forgotten an information in my first post (and which is explained in the Notepad++ user manual) : to modify the file overrideMap.xml which is located as well in the folder functionList. At the end of this file one can add a line to describe the user defined languages, here one can add the ODF language as this : 

			
			<association id= "odf.xml"				userDefinedLangName="ODF"/>
			
		</associationMap>
	</functionList>
</NotepadPlus>

I think that it is important to give the same language name (here "ODF") between the field userDefinedLangName and what is defined in the languages dialog box (where the file "Notepad++ Language ODF.xml" has to be used).

Thus I hope that you will get your ODF functions list in the application.

**Reply by jason-barba** [reply]
Thanks. That last remark did the trick!! :-)

I removed Notepad++ completly from my computer and re-installed the latest version. I did not ask for the portable version, nor did I check the portable installation box, evenso the functionList folder ended up in the executable folder, not in my appdata folder. Anyway, copying your function xml into the **functionList** folder and finally adding the above line to **overrideMap.xml** finished the work with success.

BTW The latest version 8.1.9.3 has the **Define your language ...** not at the end of the drop down menu **Language** but rather under it's submenu **User Defined Language**.

**Reply by paolo2504** [reply]
I’m having same problems… notepad as installed application, not portable… function file renamed and saved under the functionalist folder in the installation directory of notepad… modified the override map file… noway: no function visible in notepad. 
the strange thing is that override map already contains three functions and I cannot see any.

**Reply by eturpault** [reply]
@paolo2504 : I suggest you to use my new application posted today in this post (for both Windows and Linux), It is more convenient than using Notepad++ to edit an ODF in plain text.

**Reply by eturpault** [reply]
Solution found for who still doesn't arrive to have the ODF function list feature working in Notepad++ : in the folder \Notepad++\functionList (where it is necessary to add the odf.xml file provided before) there is a file named overrideMap.xml. Inside this file it is necessary to add the following line (other user language definition lines should be already present) at the end of the file, in the section User Defined Languages :

association id= "odf.xml"              userDefinedLangName="ODF"

The language ODF has to have been added as described in a previous comment.

**Reply by jason-barba** [reply]
Have to re-open this thread.
Imported the language .xml file ok.
Copied the function .xml file to functionList folder and added the association clause to overideMap.xml.

			
			<association id= "krl.xml"				userDefinedLangName="KRL"/>
			<association id= "sinumerik.xml"		userDefinedLangName="Sinumerik"/>
			<association id= "universe_basic.xml"	userDefinedLangName="UniVerse BASIC"/>
			<association id= "odf.xml"				userDefinedLangName="ODF"/>
			<!-- ======================================================================== 

Restarted Notepad++.
But, the Function List is empty :-(
Please, advice me.

**Reply by eturpault** [reply]
I repeated again on my computer (with a fresh new installed Notepad++) the installation procedure (import .xml language file in Notepad++, copied funtion .xml file in the functionList folder, added its reference in the overideMap.xml file), I do have the sections names visible in the function list of Notepad++.
Maybe you have two Notepad++ installation folders on your computer, for example (if you are in Windows) one in C:\Users\MyName\AppData\Roaming\Notepad++ and one in C:\Program Files\Notepad++, and you edited overideMap.xml in the folder which is not used by Notepad++, and you placed the function .xml in a functionList folder which is not used by Notepad++.

Else you can use OdfEdit which is more convenient than Notepad++ to edit an ODF.

**Reply by jason-barba** [reply]
You were right!
My Notepad++ had its installation folder also in C:\Users\MyName\AppData\Roaming\Notepad++. So, I created a functionList sub- folder in that folder (there was none before) and copied the function .xml to it and renamed it. And success!

Thanks for kind help.

BTW I tried to install OdfExit on an Ubuntu 22.04 LTS host. After installing pillow etc I stopped at installing PortAudio. I couldn't find a reasonable installation guide for that :-(

**Reply by eturpault** [reply]
About PortAudio installation issue in Linux, here is a solution which worked for another user : https://github.com/GrandOrgue/OdfEdit/issues/42

Note : it is OdfEdit, not Odf**Exit** (funny typo) :wink:

**Comment by Jean-Gabriel1982** [comment]
Wahou !
Thank you very much @eturpault for this !

Is that work on Linux ?

Is there a way to make an explanation of the structure of the ODF file ?
I try to create an ODF file but I don't understand at all the structure, the requested options etc ...

**Reply by eturpault** [reply]
Hello Jean-Gabriel,
Nice to cross you in this forum (last time was in Mérigny I think).
Notepad++ is not built for Linux, but I found a tutorial explaining how to install it all the same : https://itsfoss.com/notepad-plus-plus-linux/. If you succeed in installing it in Linux, the text colouring and functions list xml that I provided should work I suppose.

About the ODF file structure, personally I learned it by looking at examples (like demo.organ provided by default with GO, or the sample sets provided by @larspalo or Piotr Grabowski) and mainly thanks to the help file : GrandOrgue-help-2019-01-04.pdf.
If you have some precise questions I can try to help you with the knowledge that I have got.

**Reply by Jean-Gabriel1982** [reply]
> Hello Jean-Gabriel,
> 
> Nice to cross you in this forum (last time was in Mérigny I think).

Yes that's right 😄 I'm very happy to see that we use the same VPO !
Maybe in the future Merigny's organ will be recorded for GO !? 

> Notepad++ is not built for Linux, but I found a tutorial explaining how to install it all the same : https://itsfoss.com/notepad-plus-plus-linux/. If you succeed in installing it in Linux, the text colouring and functions list xml that I provided should work I suppose.
> 
Ok I'll try to install it 
> 
> About the ODF file structure, personally I learned it by looking at examples (like demo.organ provided by default with GO, or the sample sets provided by @larspalo or Piotr Grabowski) and mainly thanks to the help file : GrandOrgue-help-2019-01-04.pdf.
> 
> If you have some precise questions I can try to help you with the knowledge that I have got.

Many thanks again, yes it is a good idea to learn with demo organ. 
My main question is about creating panels for two screens, and adapting the resolution to screens

**Reply by eturpault** [reply]
I can precisely help in creating panels, it is was I did with the excellent Friesach organ of Piotr Grabowski, and using the Excel macros that I provided in this discussion. I added one custom panel with all the stops transtated in French (it should not be a problem for you :wink:) and the generals/sequencer in the same panel. I used the new ODF panel format for that. I can provide you my .organ file and the associated image resources if you want.
Here is the result : 
![Friesach - Mon panneau](https://user-images.githubusercontent.com/51839171/147560133-42acd4c4-9f20-4fad-8927-e0eb9435c85c.png)

**Reply by Jean-Gabriel1982** [reply]
> I can precisely help in creating panels, it is was I did with the excellent Friesach organ of Piotr Grabowski, and using the Excel macros that I provided in this discussion. I added one custom panel with all the stops transtated in French (it should not be a problem for you :wink:) and the generals/sequencer in the same panel. I used the new ODF panel format for that. I can provide you my .organ file and the associated image resources if you want.
> 
> Here is the result : 
> 
> ![Friesach - Mon panneau](https://user-images.githubusercontent.com/51839171/147560133-42acd4c4-9f20-4fad-8927-e0eb9435c85c.png)
> 
> 
> 
> 

Wahou ... so nice rendering !! It is beautiful ! Yes thank you very much I'm very interested in having your .organ file 😊 

I usually use Caen sound bank by Ben Carter but used Friesach and St John Cantius before. 

My goal is to do my own composition, adding stops, creating new (septieme, quinte 10 2/3 etc from existing stops etc ..)

And I don't really understand the structure of ODF files. 
I'll work with the demo organ. 

Regarding panels one problem is creating for one screen, the second is creating for two screens with right resolution and good position of stops. 

It is a nice idea to add the sequencer in the same panel !!

**Reply by eturpault** [reply]
Here are the files to obtain this additional panel with the Friesach sample set of Piotr Grabowski : [MonFriesach.zip](https://github.com/GrandOrgue/grandorgue/files/7786329/MonFriesach.zip)
The file "MonFriesach.organ" and the folder "Data - Friesach" have to be put in the folder where is located the original file Friesach.organ.
It is designed for a screen having a 720p vertical resolution.
We can have a call so that I explain you how I did it (I will contact you by LinkedIn).

**Reply by Psi58** [reply]
[@eterpault](https://github.com/eturpault) thanks for your contributions - I'm using the Notepad++ extensions and also ODFEdit which are both very helpful for a newbie. You mention "the Excel macros that I provided in this discussion.." - but I can't find them, could you provide a link? 
For my first "project" I'm trying to add extra panels to an existing sampleset, and it sounds like these could be useful.

**Reply by eturpault** [reply]
The Excel macros were the embryo of what OdfEdit is now, it is better that you use OdfEdit which has very much more features.
For adding sections in an ODF, OdfEdit has a "Add" button for that, permitting to link automatically the added section to its parent section.
If you want to use Excel all the same, here is the file :
[GrandOrgue ODF Excel editor.zip](https://github.com/GrandOrgue/grandorgue/files/12542380/GrandOrgue.ODF.Excel.editor.zip)

**Reply by larspalo** [reply]
@Psi58 And a GUI editor for GrandOrgue odfs that allow creating additional panels and much more can be found at https://github.com/larspalo/GOODF if you want to avoid digging in the text code.

**Reply by Psi58** [reply]
Thank you - this looks very promising for enthusiasts who are not programmers (ie people like me...) 😉

**Comment by eturpault** [comment]
Hello,
For who is interrested, here is an application (that I called ODFedit) which permits to edit an ODF in plain text, with navigation among the objects sections by list or tree, and with an integrated help and syntax / data check.
I implemented it in Python, so it can run on Windows, Linux, Mac.
If you test it, please inform me in this messages thread if you see bugs or errors, and if you have suggestions of improvement.
Here is a screenshot of the application:
![ODFedit screenshot](https://user-images.githubusercontent.com/51839171/163623745-4605a038-171c-47fc-a456-29f78993e401.png)
It is available in the zip with the source code (Python3 script) and the executable files for Windows and Linux.

**Reply by eturpault** [reply]
Update with the version v1.3.
Change log :
- data management improvement, 
- change in the way to define the parent-child relations between the objects,
- completed some attributes values maximum check, 
- added a tab to search a string in the whole ODF

[ODFedit_v1.3.zip](https://github.com/GrandOrgue/grandorgue/files/8733792/ODFedit_v1.3.zip)

@oleg68 : would you agree that in the files repository of GO a folder "Tools" is added, and inside it a folder "ODFedit" in which I could put the last version of my tool ? This would give it more visibility and may help some people who want to edit their ODF in plain-text and check its content vs the ODF format specification.

**Comment by oleg68** [comment]
@eturpault I don't object to include ODFEdit to the GrandOrgue distribution, but it needs to work more

1. To solve the dependenciy problems. For now it can not be easy run under linux:

```
[oleg@oleg2 Linux]$ ./ODFedit 
[683436] Error loading Python lib '/tmp/_MEIC69m7r/libpython3.10.so.1.0': dlopen: /lib64/libm.so.6: version `GLIBC_2.35' not found (required by /tmp/_MEIC69m7r/libpython3.10.so.1.0)
```

2. To make building from the source, including adding OdfEdit as a git submodule to the GrandOrgue source code
3. To rename ODFedit to OdfEdit according to the naming conventions of GrandOrgue

**Reply by eturpault** [reply]
@oleg68 thanks for your positive feedback.

1. I tested my tool in a Linux virtual machine, both the script (OdfEdit.py) and the executable (OdfEdit), using Ubuntu 18.04 where Python 3.10.4 is installed, and it works well. Maybe the issue that you observe is caused by the fact that an older version of Python is installed on your machine. But this is unexpected because the executable (generated with PyInstaller) is supposed to include all what is needed to run autonomously without Python installed in the machine. It appears to not be true.
The best way to run this application is to run it as a Python script (so from the source .py file) and not as an executable, so by launching it in the Linux console (having Python installed in the machine of course) with this command : 
    python3 <folder where is the script and his resources sub-folder>/OdfEdit.py

2. I need to learn how to use git, then I can do this.

3. no problem to name it OdfEdit.

**Reply by eturpault** [reply]
@oleg68 
I am ready to commit my tool OdfEdit in GitHub (in a new folder named tools\OdfEdit\), but I have not the write access to the grandorgue repository. Could you kindly grant it to me ? Thank you.

**Reply by oleg68** [reply]
@eturpault You may create a pull request with your tool.

But it is necessary to provide a valid building and packaging scripts for your tool BEFORE it may be merged. Now GrandOrgue use cmake as a build system. So you should provide cmake scripts for the integration with GrandOrgue

**Reply by eturpault** [reply]
Are building and packaging scripts really needed for a tool which has not to be build with GrandOrgue or integrated in it, and which intends to be used appart from GrandOrgue for ODF files editing ?
I had in mind simply to push the files of the tool in a directory "Tools" in the GrandOrgue repository and that's it. If it has to be a more complex operation so I forget this idea.

**Reply by oleg68** [reply]
GrandOrgue users mostly do not retrieve the souce code. They usually download binary packages from https://github.com/GrandOrgue/grandorgue/releases and install them. So they need to have this tool in the binary packages, otherwise they can not get any benefit from OdfEdit.py in the GrandOrgue source code repository

Usually a cmake script is responsible for placing files in the binary packages even the files are just copied and not compiled.

If you want to just store the source code of OdfEdit without integrating it to the GrandOrgue build and distribution system, I think creating a separate repository in Github for OdfEdit would be a better option. In this case a reference to the separate OdfEdit repository may be placed to the GrandOrgue Wiki.

**Reply by eturpault** [reply]
Thanks @oleg68 for your suggestions. I think the idea of having a dedicated repository in GitHub is the right one. I have just created this repo, it is https://github.com/eturpault/OdfEdit.
Could you kinkly reference it in the GrandOrgue Wiki ? Thank you so much !

**Reply by oleg68** [reply]
@eturpault  I've added a reference to OdfEdit on the bottom of the GrandOrgue FAQ https://github.com/GrandOrgue/grandorgue/wiki/FAQ

**Reply by rousseldenis** [reply]
@eturpault @oleg68 Maybe this could be in its own repo under GrandOrgue organization ?

**Reply by rousseldenis** [reply]
As @eturpault could be the maintainer, he could have write access on that repo in particular

**Reply by oleg68** [reply]
@rousseldenis I agree.

**Reply by rousseldenis** [reply]
Sorry not having be more active these weeks, but I've just realized I was following just the 'mention' threads.. I'll try to get back into conversations.

If we agree, I can create the repository. @larspalo

**Reply by larspalo** [reply]
@rousseldenis Sure, I have no problem with that. I've sadly had far too little time to work on GO or related projects myself, but hopefully that will be better in the future.

**Reply by eturpault** [reply]
@rousseldenis I like your suggestion, so the tool would stay "at home" close to GrandOrgue. If it is ok for the others, please let me know the name of the repository that you will have created. Thanks !

**Reply by rousseldenis** [reply]
Hi guys,

I've created this repository:

https://github.com/GrandOrgue/ODFEdit

@eturpault We'll see you there with a first PR :smiley:

**Reply by eturpault** [reply]
Hi @rousseldenis, Thanks for the creation of this repo. I think that @oleg68 would prefer that you name it OdfEdit instead of ODFEdit as he wrote in this discussion on May 22nd.

**Reply by rousseldenis** [reply]
> Hi @rousseldenis, Thanks for the creation of this repo. I think that @oleg68 would prefer that you name it OdfEdit instead of ODFEdit as he wrote in this discussion on May 22nd.

Indeed, but I think he refers more to the tool name itself than repo name(we have GrandOrgue in grandorgue repo).

So, as communication was already made on that, let's go like this and name our tool OdfEdit.

**Reply by eturpault** [reply]
I have populated the repo GrandOrgue/ODFEdit with my files.
@rousseldenis  : is it possible to enable the Discussions feature for this repo ? It could be used if somebody wants to suggest/discuss evolutions of the tool.
@oleg68 : could you please change the link in the WiKi FAQ to this new repo ? I will then delete my repo eturpault/OdfEdit. Thank you.

**Reply by rousseldenis** [reply]
@eturpault Thanks for this.

The idea was to create a Pull request in order to review your code.

We we'll keep it as is, but for next changes, it's better to use that flow.

I can enable Discussions. For features/bug requests, creating issues is better.

**Reply by eturpault** [reply]
Sorry I am new in Git, I will do pull request on next times.
If issues is better for features requests as well, so no need to enable Discussions.

**Reply by oleg68** [reply]
> could you please change the link in the WiKi FAQ to this new repo ? I will then delete my repo eturpault/OdfEdit. Thank you.

Done.
